### PR TITLE
add 1-worker bases + example

### PR DESCRIPTION
**Pull request by jnunyez**
- Bases for adding a worker node
- example to use adding a worker node

### CODE REVIEW

File: overlay/1-worker/patches/bmh-volante-cars-lab-du3-ldc1.yaml [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+  value: redfish-virtualmedia://172.28.11.44/redfish/v1/Systems/1/
+- op: replace
+  path: /spec/bmc/credentialsName
+  value: bmc-du3-ldc1-volante-cars-lab
+- op: replace
+  path: /spec/bmc/disableCertificateVerification
+  value: true
+- op: replace
+  path: /metadata/labels/infraenvs.agent-install.openshift.io
+  value: infraenv
```
Comment (natifridman) [inline overlay/1-worker/patches/bmh-volante-cars-lab-du3-ldc1.yaml]: Is this the infraenv name?

**Review by natifridman — COMMENTED** [review]
Looks good, I would also add a readme explaining that this is for an existing cluster and what is required to add a worker

File: overlay/1-worker/patches/bmh-volante-cars-lab-du3-ldc1.yaml (hunk 1 above)
Comment (jnunyez) [reply overlay/1-worker/patches/bmh-volante-cars-lab-du3-ldc1.yaml]: yes, that was it.

**Comment by jnunyez** [comment]
@natifridman @novacain1  what do you think of nmstateconfig for a worker ndoe being fully declared in the overlay with no base? Is there any part you think than ca be included in the base? Currently all the nmstateconfig is located in the overlay. 

[nmstateconfig](https://github.com/jnunyez/vse-ocp-bases/blob/add-worker/overlay/1-worker/patches/nmstate-config.yaml)

**Comment by natifridman** [comment]
> @natifridman @novacain1 what do you think of nmstateconfig for a worker ndoe being fully declared in the overlay with no base? Is there any part you think than ca be included in the base? Currently all the nmstateconfig is located in the overlay.
> 
> [nmstateconfig](https://github.com/jnunyez/vse-ocp-bases/blob/add-worker/overlay/1-worker/patches/nmstate-config.yaml)

As I see it, you should include the `NMStateConfig` for the Sync Waves. The customer/user doesn't really care/know in which wave this should be included.
So you can add the minimum required manifest in the base as we did for the [3-node](https://github.com/redhat-partner-solutions/vse-ocp-bases/blob/main/base/3-node/nmstate-config.yaml) and use `patchesStrategicMerge` in the overlay for the patches

**Comment by novacain1** [comment]
The most basic config that can be included is IMHO a VM with a single interface being used with dual stack, static IPv4/IPv6.  The possibilities with NMState are quite vast, especially with interface channel bonding, bridges and the like, etc.

IMHO we should default to static IPs for interfaces we care about vs. relaying on the default behavior, which is DHCP on all interfaces.

**Comment by jnunyez** [comment]
> As I see it, you should include the `NMStateConfig` for the Sync Waves. The customer/user doesn't really care/know in which wave this should be included. So you can add the minimum required manifest in the base as we did for the [3-node](https://github.com/redhat-partner-solutions/vse-ocp-bases/blob/main/base/3-node/nmstate-config.yaml) and use `patchesStrategicMerge` in the overlay for the patches

Note also one difference w.r.t  the 3-node case is that it is well known a priori the number of objects you will have (i.e., 3) meaning that you can set the name and number of the NMstateConfig objects. In the case of the day-2 workers it is not known the number of workers the user will have to add so the user cannot try to reuse one single base NMState object to try to provision two worker nodes. If the user wants to add two worker nodes, the user will have to add  an NMStateConfig to its base yaml.

File: overlay/1-worker/patches/bmh-volante-cars-lab-du3-ldc1.yaml (hunk 1 above)
Comment (jnunyez) [reply overlay/1-worker/patches/bmh-volante-cars-lab-du3-ldc1.yaml]: Added a little description [here](https://github.com/jnunyez/vse-ocp-bases/blob/add-worker/base/1-worker/README.md)

**Comment by jnunyez** [comment]
> The most basic config that can be included is IMHO a VM with a single interface being used with dual stack, static IPv4/IPv6. The possibilities with NMState are quite vast, especially with interface channel bonding, bridges and the like, etc.
> 
> IMHO we should default to static IPs for interfaces we care about vs. relaying on the default behavior, which is DHCP on all interfaces.

@novacain1  Let's keep it simple for the moment. Moved the just to keep the ArgoCD annotation overlay agnostic. Also take a look at the default proposed for provisioning object [here](https://github.com/jnunyez/vse-ocp-bases/blob/add-worker/base/1-worker/provisioning.yaml)

**Comment by natifridman** [comment]
> Note also one difference w.r.t the 3-node case is that it is well known a priori the number of objects you will have (i.e., 3) meaning that you can set the name and number of the NMstateConfig objects. In the case of the day-2 workers it is not known the number of workers the user will have to add so the user cannot try to reuse one single base NMState object to try to provision two worker nodes. If the user wants to add two worker nodes, the user will have to add an NMStateConfig to its base yaml.

For adding 2 nodes, would you use the same overlay or create 2 different overlays for each node?

**Comment by jnunyez** [comment]
I would prefer to use the same base with a different overlay per-worker. But to do this, the name of the object NMStateConfig specified in the base folder has to change per overlay folder. Wonder if there is a way to do that.

**Comment by jnunyez** [comment]
I would prefer to use the same base with a different overlay per-worker. But to do this, the name of the object NMStateConfig specified in the base folder has to change per overlay folder. Wonder if there is a way to do that.

**Comment by jnunyez** [comment]
> > Note also one difference w.r.t the 3-node case is that it is well known a priori the number of objects you will have (i.e., 3) meaning that you can set the name and number of the NMstateConfig objects. In the case of the day-2 workers it is not known the number of workers the user will have to add so the user cannot try to reuse one single base NMState object to try to provision two worker nodes. If the user wants to add two worker nodes, the user will have to add an NMStateConfig to its base yaml.
> 
> For adding 2 nodes, would you use the same overlay or create 2 different overlays for each node?

I would like to create 2 different overlays and ideally not change the base folder. Wonder if there is a way to change NMStateConfig name object per overlay because we are defining the name in the base folder and doing a strategicMerge in the overlay that basically looks for a specific name of the NMStateConfig object to apply the patch. 
Maybe with nameSuffix or namePrefix?

**Comment by natifridman** [comment]
I thought about something like this:
Using a different folder for each node we would like to add
```
├── add_nodes
  └── overlays
      ├── node1
      │   ├── patches
      │   └── kustomization.yaml
      ├── node2
      │   ├── patches
      │   ├── kustomization.yaml
```

WDYT?

**Comment by jnunyez** [comment]
> 


> I thought about something like this: Using a different folder for each node we would like to add
> 
> ```
> ├── add_nodes
>   └── overlays
>       ├── node1
>       │   ├── patches
>       │   └── kustomization.yaml
>       ├── node2
>       │   ├── patches
>       │   ├── kustomization.yaml
> ```
> 
> WDYT?

Yes, that is the goal.  After our chat, I give it a try. The **patcheStrategicMerge** patch is applied **before** the patch replace. It is good because as  long as we dont execute the addition of the two nodes in the same `kustomize` process we can totally leverage the 1-worker base. I added an example of adding two different workers in different kustomize in [here](https://github.com/redhat-partner-solutions/vse-ocp-bases/pull/6/commits/71aab83623f8459c010e7e439d813558da9c2e73) 
We cannot apply them in pararell meaning that we cannot get a kustomize in folder `add-workers` calling 1st-worker and 2nd-worker.

**Comment by natifridman** [comment]
> Yes, that is the goal. After our chat, I give it a try. The **patcheStrategicMerge** patch is applied **before** the patch replace. It is good because as long as we dont execute the addition of the two nodes in the same `kustomize` process we can totally leverage the 1-worker base. I added an example of adding two different workers in different kustomize in [here](https://github.com/redhat-partner-solutions/vse-ocp-bases/pull/6/commits/71aab83623f8459c010e7e439d813558da9c2e73)
> We cannot apply them in pararell meaning that we cannot get a kustomize in folder `add-workers` calling 1st-worker and 2nd-worker.

I think that's ok, so in our use case when we use `applicationset` for the clusters, we can put the nodes folders in the `clusters` folder. 
This will create a different Argo CD Application for every folder (and a different kustomize build)
```
├── hub
  └── clusters
      ├── cluster
      │   ├── patches
      │   └── kustomization.yaml
      ├── node1
      │   ├── patches
      │   └── kustomization.yaml
      ├── node2
      │   ├── patches
      │   ├── kustomization.yaml
```

**Review by natifridman — APPROVED** [review]

**Pull request merged**